### PR TITLE
Writing GeoTIFFs

### PR DESCRIPTION
**Pull request by metasim**

### CODE REVIEW

File: pyrasterframes/src/main/python/docs/raster-write.pymd [hunk 1]
```diff
@@ -48,21 +50,47 @@ pandas_df.dtypes
...
-It is oftentimes convenient to have a simple overview image of the data, particularly if it spans a very wide area.
 
 ## GeoTIFFs
 
-GeoTIFF is one of the most common file formats for spatial data. 
+GeoTIFF is one of the most common file formats for spatial data, providing flexibility in data encoding, representation, and storage. RasterFrames provides a specialized Spark DataFrame writer for rendering a RasterFrame to a GeoTiff.
+
+One downside to GeoTIFF is that it is not a big data native format. To create a GeoTIFF all the data to be encoded has to be in the memory of one compute node (in Spark parlance, this is a "collect"), limiting it's maximum size substantially compared to that of a full cluster environment. When rendering GeoTIFFs in RasterFrames, you either need to specify the dimensions of the output raster, or be aware of how big the collected data will end up being.
+
+Fortunately, we can use the cluster computing capability to downlample the data into a more manageble saze. For sake of example, let's render a simple RGB overview image of our scene as a small raster:
```
Comment (vpipkt) [inline pyrasterframes/src/main/python/docs/raster-write.pymd]: spelling "saze".

File: pyrasterframes/src/main/python/docs/raster-write.pymd [hunk 2]
```diff
@@ -48,21 +50,47 @@ pandas_df.dtypes
...
+
+```python write_geotiff
+import os.path
+from docs import docs_dir
+cat = """  
+red,green,blue
+{},{},{}
+""".format(scene('01'), scene('04'), scene('03'))
+outfile = os.path.join(docs_dir(), 'geotiff-overview.tif')
+rf = spark.read.raster(catalog=cat, catalog_col_names=['red', 'green', 'blue'])
```
Comment (vpipkt) [inline pyrasterframes/src/main/python/docs/raster-write.pymd]: would be nice to show how large the original raster is.

File: pyrasterframes/src/main/python/docs/raster-write.pymd (hunk 1 above)
Comment (vpipkt) [reply pyrasterframes/src/main/python/docs/raster-write.pymd]: and can we mention the downsampling method? I assume it's nearest nieghbor? Is it configurable?

File: pyrasterframes/src/main/python/docs/raster-write.pymd [hunk 3]
```diff
@@ -48,21 +50,47 @@ pandas_df.dtypes
...
+rf.write.geotiff(outfile, crs='EPSG:4326', raster_dimensions=(256, 256))
```
Comment (vpipkt) [inline pyrasterframes/src/main/python/docs/raster-write.pymd]: worth pointing out above that we are also reprojecting the raster here.

File: pyrasterframes/src/main/python/docs/raster-write.pymd [hunk 4]
```diff
@@ -48,21 +50,47 @@ pandas_df.dtypes
...
+import rasterio
+import numpy as np
+from rasterio.plot import show, show_hist
+ 
+with rasterio.open(outfile) as src:
+    cells = np.clip(src.read(), 0, 1800).astype('float32')
+    show(cells)
+    show_hist(src, bins=50, lw=0.0, stacked=False, alpha=0.3,
+        histtype='stepfilled', title="Overview Histogram")
+```
```
Comment (vpipkt) [inline pyrasterframes/src/main/python/docs/raster-write.pymd]: I would also like to see an example where we select a contiguous subset and write it out at full resolution and in native CRS. 

Just for completeness. 

I think that would be a quite common use case also. Consider for example writing out ML classification model predictions for an area intersecting a polygon.

File: pyrasterframes/src/main/python/pyrasterframes/__init__.py [hunk 5]
```diff
@@ -167,10 +173,32 @@ def temp_name():
...
+    df_writer,
+    path=None,
+    crs=None,
+    raster_dimensions=None,
+    **options):
+
+    if raster_dimensions is not None:
+        if isinstance(raster_dimensions, tuple):
+            options.update({
+                "imageWidth": raster_dimensions[0],
```
Comment (vpipkt) [inline pyrasterframes/src/main/python/pyrasterframes/__init__.py]: may make sense to check that the tuple length is 2 and that the members are `int` type?

File: pyrasterframes/src/main/python/pyrasterframes/__init__.py [hunk 6]
```diff
@@ -167,10 +173,32 @@ def temp_name():
...
+                "imageHeight": raster_dimensions[1]
+            })
+        elif isinstance(raster_dimensions, str):
+            parts = raster_dimensions.split(',')
+            options.update({
+                "imageWidth": parts[0],
```
Comment (vpipkt) [inline pyrasterframes/src/main/python/pyrasterframes/__init__.py]: again maybe makes sense to check `len(parts)==2` and `all([isinstance(p, int) for p in parts])`

File: pyrasterframes/src/main/python/pyrasterframes/__init__.py [hunk 7]
```diff
@@ -167,10 +173,32 @@ def temp_name():
...
+                "imageHeight": parts[1]
+            })
+
+    if crs is not None:
```
Comment (vpipkt) [inline pyrasterframes/src/main/python/pyrasterframes/__init__.py]: in our discussions before i thought crs was required?

Comment (vpipkt) [reply pyrasterframes/src/main/python/pyrasterframes/__init__.py]: Yea. 

```
Py4JJavaError: An error occurred while calling o107.save.
: java.lang.IllegalArgumentException: requirement failed: A destination CRS must be provided
	at scala.Predef$.require(Predef.scala:224)
```

So we could have a check like this in place of the if statement. 

```python
```suggestion
    assert(crs is not None, "You must pass in a `crs` argument")
```

 WOuld be good in docs to give some details about what's valid. EPSG codes? proj strings? wkt? What about pyproj [CRS object](https://pyproj4.github.io/pyproj/stable/api/crs.html)?  rasterio [CRS](https://rasterio.readthedocs.io/en/stable/api/rasterio.crs.html#rasterio.crs.CRS) Possibly some of those can come later in a future issue / PR.

**Comment by vpipkt** [comment]
Excited that my initial test is looking very good!

**Comment by vpipkt** [comment]
One interesting thing. 

I read and wrote to gtiff a single MODIS granule (single channel). In the resulting `gdalinfo` the pixel size is (460.720950959172399,-460.720950959172399) 

But the original granule's pixel size is slightly larger: (463.312716527916905,-463.312716527916677)

See the movie attached, you can kind of see an artifact that may be the tile boundaries.

Happy to share original and written file.


[ScreenRecordModis.mov.zip](https://github.com/locationtech/rasterframes/files/3448602/ScreenRecordModis.mov.zip)


The tile dimensions are pretty much normal for this kind of file, so pretty irregular. The most common one is not even square.

```
+------------------+-----+
|rf_dimensions(B01)|count|
+------------------+-----+
|         [256, 96]|    6|
|        [257, 257]|   18|
|         [257, 96]|    3|
|          [97, 96]|    1|
|         [97, 257]|    6|
|        [256, 257]|   36|
|        [257, 256]|    9|
|        [256, 256]|   18|
|         [97, 256]|    3|
+------------------+-----+
```

File: pyrasterframes/src/main/python/pyrasterframes/__init__.py (hunk 5 above)
Comment (metasim) [reply pyrasterframes/src/main/python/pyrasterframes/__init__.py]: How do you do that? Nested `isinstance` calls?

Comment (metasim) [reply pyrasterframes/src/main/python/pyrasterframes/__init__.py]: Heh... nevermind.

File: pyrasterframes/src/main/python/pyrasterframes/__init__.py (hunk 7 above)
Comment (metasim) [reply pyrasterframes/src/main/python/pyrasterframes/__init__.py]: > in our discussions before i thought crs was required?

The only case where it isn't is if it's already a "layer".... probably not the regular use case however.

File: pyrasterframes/src/main/python/docs/raster-write.pymd (hunk 1 above)
Comment (metasim) [reply pyrasterframes/src/main/python/docs/raster-write.pymd]: Will add statement on method used. Making it configurable:
https://github.com/locationtech/rasterframes/issues/222

File: pyrasterframes/src/main/python/docs/raster-write.pymd (hunk 2 above)
Comment (metasim) [reply pyrasterframes/src/main/python/docs/raster-write.pymd]: https://github.com/locationtech/rasterframes/issues/223

File: pyrasterframes/src/main/python/docs/raster-write.pymd (hunk 4 above)
Comment (metasim) [reply pyrasterframes/src/main/python/docs/raster-write.pymd]: https://github.com/locationtech/rasterframes/issues/228

**Pull request merged**